### ISSUE TITLE
[SPEC] Improve usability issues on touch/mobile devices

### ISSUE THREAD

**Issue by LukasMasuch** [glideapps/glide-data-grid#194]
# Glide Data Grid Touch Specification

Goal: Every major interaction in Glide Data Grid should be touch friendly. This may be provided simply by expanding the data grid to work with touch the same it works with a mouse, or by adding alternative interaction mechanisms.

## Scroll

Scrolling should work via the standard browser swipe scrolling mechanism. There are some differences with how this should work with mouse. 

### Special considerations
- Touch move events should not be translated into mouse motion events as these are already overloaded for scrolling.
  - This even applies when no scrolling is possible as it sets up an incorrect expectation.
- Cell selection must happen on `TouchUp` and not `TouchDown`. If it happens on the down event cells will be selected as you scroll. Further `TouchUp` must not select a cell if scrolling has occurred, even if the `TouchDown` and `TouchUp` happen on the same cell (as they should).

## Resizing columns

Resizing columns can be implemented as a direct interaction, however this would interfere somewhat with standard scroll behavior. Instead when a column header is selected via a touch event, a drag thumb should appear over the right hand gutter. This drag handle can then be moved to resize the column.

## Unselect current selection

This is only not possible when `rowMarkers` are not enabled. When `rowMarkers` are enabled click the top left header cell twice will clear the selection. While this is only a partial solution, it is a majority solution and the cell is visible 100% of the time. We should document this solution better.

Future considerations: Haptic touch interaction for de-selection, this would need to use the browser api's for sending the haptic feedback. Further we would need to consider if there are other times when this style of de-select should be used.

## Multi-column/cell selection

As selection no longer happens on touch down but instead on touch up we can be a little clever here.

Multi-select flow:
- User taps on header A
- User scrolls right and then triggers `TouchDown` on header G
- If ~500ms elapse and no `TouchUp`, cancel, or scroll event is received. All columns between A and G become selected.
- If < 500ms elapse header G is set as the sole selected header.

The same flow is used for groups and cell range selection.

## Unable to trigger search

Document the need to add a search button.

## Rearrange columns/rows

Drag interactions on mobile are unlikely to be pleasant. Initially I thought long press then drag might work, but this is unlikely to be intuitive. More thought is required.

## Unable to trigger menus

What a header is activated via a touch event, it should always act as if the header is hovered. This makes the menu click area visible to the user. The click region should also be enlarged.

## Unable to trigger Cell/Header/Group context menus

Long press is the obvious solution here, with the caveat that it shouldn't do this if it activates the multi select. 

# Original Report

I did some testing today on the usability of the glide data grid on touch/mobile devices (with Chrome). Overall it works pretty well, but there are some issues I have found:

1. Touch events during scrolling change the cell selection.
2. Resizing columns does not work.
3. No way to unselect a cell on mobile (since `ESC` hotkey is not possible). Possible touch interaction: Long-press on selected cell unselects it.
4. Multi-column selection is not possible.
5. Unable to trigger the search (since `CTRL` + `F` hotkey is not possible)
6. Reranging columns does not work.
7. Reordering Rows does not work;

I think 1 has the most impact on improving the usability on touch devices in case cell selection triggers something within an application. 2-7 are probably more nice to have features regarding touch usability.

**Comment by ivoelbert** [comment]
Thanks for submitting this!

I agree, `1.` is a major pain point. We'll work on this soon.

Regarding
> Unable to trigger the search (since CTRL + F hotkey is not possible)

You can show/hide search by using the `showSearch` prop. `ctrl`+`f` is just convenient in desktop.
Here's a story: https://glideapps.github.io/glide-data-grid/?path=/story/dataeditor--built-in-search

**Comment by jassmith** [comment]
Proposed fixes

## Touch events during scroll change the cell selection

Selection on mobile should be based on touch end events and should only select if less than 20px of scrolling has ocured in either direction.

## Resizing columns does not work

Pinch to zoom on column header maybe? Would be hard to resize small columns. Tap and drag on gutters would also work.

## No way to unselect a cell

I kind of think this might be better as an application level problem. Clear cell button, long press, there are options. Anyway I think this is out of scope.

## Multi-column selection

Proposed interaction: long press on column header behaves the same as a shift+click

## Unable to trigger search on mobile

Application problem out of scope. A button needs to be added to the application UI around the chrome.

## Rearanging rows/columns does not work

Long press then drag seems optimal here. Long press and drag on cell maybe could enable drag select also.

**Comment by LukasMasuch** [comment]
Thanks for the fast responses! Agreed on the proposed fixes 👍  a few further comments:

**Resizing columns does not work**

I think tap & drag is fine if the touch target is big enough. I also did a quick check on how some other table components do the resizing on mobile: [MUI Data Grid](https://mui.com/components/data-grid/columns/#resizing) & [Ag-grid](https://www.ag-grid.com/javascript-data-grid/column-sizing/#resizing-example) just have tap & drag on the header border. Notion tables seem to not allow resizing on touch devices. And Google Sheets has this resizing dot within the mobile app:

<img src="https://user-images.githubusercontent.com/2852129/157491565-56d268ce-39cb-499f-9c66-67061c7a2c79.png" width="250">

**Rearranging rows/columns does not work**

> Long press and drag on cell maybe could enable drag select also.

I think long-press on a cell might be needed for the right click (`onCellContextMenu`)? Maybe this is actually an additional usability issue to add to the list:

8. Context menus (`onCellContextMenu`, `onGroupHeaderContextMenu`, `onHeaderContextMenu`) are not triggerable on touch devices  At least, for the cell context menu this might be solvable via long-press. For the header context menu, this might interfere with the multi-column selection - in case this uses long-press as well.

**Comment by jassmith** [comment]
I've updated the issue description with a spec

**Comment by jassmith** [comment]
I've fixed all of these except for resizing/dragging. I am going to close this for now. We can revisit those later.
